source/utils/body.ts

```diff
@@ -54,15 +54,43 @@ export const getBodySize = (body?: BodyInit | null): number => {
 	return 0; // Default case, unable to determine size
 };
 
-export const streamResponse = (response: Response, onDownloadProgress: Options['onDownloadProgress']) => {
-	const totalBytes = Number(response.headers.get('content-length')) || 0;
+const withProgress = (stream: ReadableStream<Uint8Array>, totalBytes: number, onProgress: Options['onDownloadProgress'] | Options['onUploadProgress']): ReadableStream<Uint8Array> => {
+	let previousChunk: Uint8Array | undefined;
 	let transferredBytes = 0;
 
-	if (response.status === 204) {
-		if (onDownloadProgress) {
-			onDownloadProgress({percent: 1, totalBytes, transferredBytes}, new Uint8Array());
-		}
+	return stream.pipeThrough(new TransformStream<Uint8Array, Uint8Array>({
+		transform(currentChunk, controller) {
+			controller.enqueue(currentChunk);
+
+			if (previousChunk) {
+				transferredBytes += previousChunk.byteLength;
+
+				let percent = totalBytes === 0 ? 0 : transferredBytes / totalBytes;
+				// Avoid reporting 100% progress before the stream is actually finished (in case totalBytes is inaccurate)
+				if (percent >= 1) {
+					percent = 1 - Number.EPSILON;
```

Comment:
```suggestion
					// Prevents percentage from going backwards when close to 100%.
					percent = 1 - Number.EPSILON;
```

Reply:
That is only true relative to the other approaches we could've taken here. I think this comment would be confusing when read by someone in the future.

Reply:
I wrote a comment above the `if` condition that I think explains the purpose pretty well. Did you see that? The choice of `Number.EPSILON` is just a way to achieve that goal. It's merely the largest number that isn't 100%.